client/state/imports/reducer.js

```diff
@@ -1,13 +1,182 @@
+import siteImporter from './site-importer/reducer';
+import { appStates } from './constants';
+import { fromApi } from './api';
+
+function isImporterStatusHydrated( state = false, action ) {
+	switch ( action.type ) {
+		case IMPORTS_IMPORT_RECEIVE:
+			return true;
+	}
+
+	return state;
+}
+
+function importerStatus( state = {}, action ) {
+	switch ( action.type ) {
+		case IMPORTS_IMPORT_CANCEL:
+		case IMPORTS_IMPORT_RESET:
+			return omit( state, action.importerId );
+
+		case IMPORTS_UPLOAD_FAILED:
+			return {
+				...state,
+				[ action.importerId ]: {
+					...state[ action.importerId ],
+					importerState: appStates.UPLOAD_FAILURE,
+					errorData: { type: 'uploadError', description: action.error },
+				},
+			};
+
+		case IMPORTS_UPLOAD_COMPLETED:
+			return {
+				...omit( state, action.importerId ),
+				[ action.importerStatus.importerId ]: action.importerStatus,
+			};
+
+		case IMPORTS_IMPORT_RECEIVE: {
+			// The REST endpoint returns an empty `[]` array if there are no known imports.
+			// In that case we don't update the `importerStatus` map, and only mark it as hydrated.
+			if ( isEmpty( action.importerStatus ) ) {
+				return state;
+			}
+
+			// don't receive the response if the importer is locked
+			if ( action.isLocked ) {
+				return state;
+			}
+
+			// convert the response with `fromApi` only after we know it's not empty
+			const newImporterStatus = fromApi( action.importerStatus );
+
+			return omitBy(
+				{
+					// filter the original set of importers...
+					...state,
+					// ...and the importer being received.
+					[ newImporterStatus.importerId ]: newImporterStatus,
+				},
+				( importer ) =>
+					[ appStates.CANCEL_PENDING, appStates.DEFUNCT, appStates.EXPIRED ].includes(
+						importer.importerState
+					)
+			);
+		}
+
+		case IMPORTS_AUTHORS_START_MAPPING:
+			return {
+				...state,
+				[ action.importerId ]: {
+					...state[ action.importerId ],
+					importerState: appStates.MAP_AUTHORS,
+				},
+			};
+
+		case IMPORTS_AUTHORS_SET_MAPPING:
+			return {
+				...state,
+				[ action.importerId ]: {
+					...state[ action.importerId ],
+					customData: {
+						...state[ action.importerId ].customData,
+						sourceAuthors: map(
+							get( state[ action.importerId ], 'customData.sourceAuthors' ),
+							( author ) =>
+								action.sourceAuthor.id === author.id
+									? {
+											...author,
+											mappedTo: action.targetAuthor,
+									  }
+									: author
+						),
+					},
+				},
+			};
+
+		case IMPORTS_UPLOAD_SET_PROGRESS:
+			return {
+				...state,
+				[ action.importerId ]: {
+					...state[ action.importerId ],
+					percentComplete: ( action.uploadLoaded / ( action.uploadTotal + Number.EPSILON ) ) * 100,
+				},
+			};
+
+		case IMPORTS_IMPORT_START:
+			return {
+				...state,
+				[ action.importerId ]: {
+					importerId: action.importerId,
+					importerState: appStates.READY_FOR_UPLOAD,
+					site: { ID: action.siteId },
```

Comment:
Would be cool if we directly store the `siteId` here if we only have an `ID` in that `site` object.

Reply:
I agree and would like to do this in a targeted followup PR. Generally, I think the mapping between the REST response format and the state format could be more 1:1, with less work for the `fromApi` and `toApi` functions.